lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -4928,36 +4934,36 @@ template <class T> bool fuseActivation(T *N, Function *F, const Backend *B) {
     return false;
   }
 
-  FusedActivation activationType;
   NodeValue activationNV;
   switch (activation->getKind()) {
   case Kinded::Kind::ReluNodeKind:
-    activationType = FusedActivation::RELU;
     activationNV = cast<ReluNode>(activation)->getResult();
+    N->setFusedActivation(FusedActivation::RELU);
+    break;
+  case Kinded::Kind::ClipNodeKind:
+    activationNV = cast<ClipNode>(activation)->getResult();
+    N->setFusedActivation(FusedActivation::CLIP);
+    N->setFusedActivationArgs({cast<ClipNode>(activation)->getMin(),
+                               cast<ClipNode>(activation)->getMax()});
     break;
   case Kinded::Kind::SigmoidNodeKind:
-    activationType = FusedActivation::SIGMOID;
     activationNV = cast<SigmoidNode>(activation)->getResult();
+    N->setFusedActivation(FusedActivation::SIGMOID);
     break;
   case Kinded::Kind::TanhNodeKind:
-    activationType = FusedActivation::TANH;
     activationNV = cast<TanhNode>(activation)->getResult();
+    N->setFusedActivation(FusedActivation::TANH);
+    break;
+  case Kinded::Kind::LeakyReluNodeKind:
+    activationNV = cast<LeakyReluNode>(activation)->getResult();
+    N->setFusedActivation(FusedActivation::LEAKY_RELU);
+    N->setFusedActivationArgs({cast<LeakyReluNode>(activation)->getAlpha()});
     break;
   default:
     return false;
   }
 
-  // currently not to support asymmetric quantization fusion.
```

Comment:
@mciprian13 Hello, I am Jemin. I wrote this code to prevent the fusion in case of asymmetric due to the offset. In asymmetric case, the offset of Conv. should be handled properly rather than replacing it by Relu's one. I wonder that you are aware of this issue.

Reply:
@leejaymin I don't understand what is the problem of fusing an asymmetric activation? The libjit correctly handles an activation with asymmetric quantization. You can look in libjit_defs.h at the function libjit_activation_q().

Reply:
@mciprian13 Sorry for the ambiguous explanation. What I concern about is quantization parameters of convolution are substituted by ones of activation.  In the symmetric case, A(partialSum) * matMul(conv) / outScale (conv) * InScale(relu) / outSacle (relu) is equal to A * matmul(conv) / outScale(relu). It is beacuse outScale (conv) is equivalent to inScale(relu). Therefore, the replacement of quantization parameters is fine. However, In the asymmetric case, there are offsets in Conv and activation. Instead of replacing them with the following code `N->getResult().setType(activationNV.getType())`, the original offset of activation should be properly computed in libjit_activation_i32(). Let me know If I was wrong.

Reply:
@leejaymin Changing the quantization parameters (scale and offset) of the convolution is not a problem because the mathematics remains consistent. Doing so actually has benefits because the dynamic range of activation output is commonly smaller than the dynamic range of the activation input (conv output). Computing the convolution using the quantization parameters of the activation:
- ensures better representation (smaller quantization step and hence error) for those values which are not saturated by Relu/Clip
- saturates those values which will be later saturated by Relu/Clip (don't care)
- this is the approach used by TensorFlowLite which ALWAYS uses fused activations in their pre-quantized models
From my experience fusing activations for quantized models always has benefits in terms of accuracy because it avoids an extra requantization (by using different quantization parameters at the interface between conv and activation).

Reply:
@mciprian13 Thank you for the really thorough answer and for sharing me the insightful document. As you explained in the attached document, I agreed with the benefit of performance due to mathematical simpleness. One thing I still concern about is that `S_y` and `O_y` are forced to match `S_a` and `O_a`. Such force conversion could affect the final accuracy. 
![image](https://user-images.githubusercontent.com/1924507/95959819-f3bd5e80-0e3d-11eb-8fe0-981dbbb0431a.png)
To figure out it, I tested accuracy variation by fusion on OpenCL backend and I got the following results:

```bash
# use imagenet12 validation set (50,000)
# baseline accuracy (FP32) top1 76.08%

# Resnet50, Asymmetric, Fusion (Conv and Relu)
Completed running; Final Top-1/5 accuracy across 50000 images:
       Top-1 accuracy: 0.11

# Resnet50, Asymmetric nonFusion
Completed running; Final Top-1/5 accuracy across 50000 images:
       Top-1 accuracy: 69.8
```

Even though the different thing between the two conditions is ON/OFF of Conv+Relu fusion, the accuracy drop is severe. However, except for the asymmetric case, the fusion slightly improves total accuracy as you mentioned. Therefore, I added the code to prevent fusion in the case of the asymmetric schema.
Currently, I am not sure that the poor accuracy stems from fusion or there might be something that is the wrong implementation in the OpenCL backend. 

In your code, I wonder if the force conversion doesn't matter.

Reply:
@leejaymin The OpenCL implementation is obviously wrong:
https://github.com/pytorch/glow/blob/7fa10073a6641ecd53b3789acfe4d1c878155ce4/lib/Backends/OpenCL/kernels.cl#L1181
The max should be done with **destOffset** and not 0 to capture the general case for asymmetric quantization.

Reply:
@leejaymin Could you try and compute again the accuracy? I fixed the OpenCL kernel by replacing 0 with destOffset.

Reply:
@mciprian13 Thank you for your quick reply. 
I pulled your PR and ran it on CPU and OpenCL backends.
Even if you fixed the openCL code, the accuracy still shows the poor result as the below.
​```bash
Finished image index 1300 out of 50000
  Current Top-1/5 accuracy:
        Top-1 accuracy: 0.00
        Top-5 accuracy: 0.00
```
In CPU backend case, the fusion of the asymmetric case shows the right result.
​```bash
Finished image index 100 out of 50000
  Current Top-1/5 accuracy:
        Top-1 accuracy: 94.00
        Top-5 accuracy: 97.00
00:00:44.30
Finished image index 200 out of 50000
  Current Top-1/5 accuracy:
        Top-1 accuracy: 80.50
        Top-5 accuracy: 97.00
```
In light of the outcome, there is extra something wrong in the OpenCL backend.
Thank you for your clarification and guidance.
I hope your code is getting landed soon.